### ISSUE TITLE
vm/ffi] DynamicLibrary executable and process have nearly identical behavior

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#44856]
We have two factory constructors in `DynamicLibrary`, `process` and `executable`, which currently have identical behavior. They can lookup symbols in the executable and statically linked libraries, but not dynamically linked libraries.(x)

They would behave differently if `DynamicLibrart.open` would load a library with global visibility instead of local visibility. In that case `DynamicLibrart.process` would be able to see the symbols, but `DynamicLibrart.executable` would not.

We could do one of the following:
1. Deprecate `DynamicLibrary.process` and always use `DynamicLibrary.executable`. (`DynamicLibrary.process` is not available on Windows, `DynamicLibrary.executable` is.)
2. Change `DynamicLibrary.open` to enable loading a library with global visibility by adding a named argument.

```dart
class DynamicLibrary {
  external static DynamicLibrary open(String path, {bool globalVisibility = false});
}
```

(x) One can do a lookup of `dlopen`, and then `dlopen` with global visibility. That would make the symbol visible in `DynamicLibrart.process`.

**Comment by dcharkes** [comment]
https://github.com/dart-lang/sdk/issues/50105#issuecomment-1441764283

MacOS and Linux open with GLOBAL, while Android does not. That explains the difference between the Flutter and Dart experience. Flutter Desktop should also work with GLOBAL.

The workaround is to call `dlopen` yourself with the right arguments

```dart

/// On Linux and Android.
const RTLD_LAZY = 0x00001;

/// On Android Arm.
const RTLD_GLOBAL_android_arm32 = 0x00002;

/// On Linux and Android Arm64.
const RTLD_GLOBAL_rest = 0x00100;

final RTLD_GLOBAL = Abi.current() == Abi.androidArm
    ? RTLD_GLOBAL_android_arm32
    : RTLD_GLOBAL_rest;

@Native<Pointer<Void> Function(Pointer<Char>, Int)>()
external Pointer<Void> dlopen(Pointer<Char> file, int mode);

/// Returns dylib
Object dlopenGlobalPlatformSpecific(String name, {String? path}) {
  if (Platform.isLinux || Platform.isAndroid || Platform.isFuchsia) {
    // TODO(https://dartbug.com/50105): enable dlopen global via package:ffi.
    return using((arena) {
      final dylibHandle = dlopen(
          platformPath(name).toNativeUtf8(allocator: arena).cast(),
          RTLD_LAZY | RTLD_GLOBAL);
      return dylibHandle;
    });
  } else {
    // The default behavior on these platforms is RLTD_GLOBAL already.
    return dlopenPlatformSpecific(name, path: path);
  }
}
```

In the future, you will be able to use `@Native` without relying on opening things global. https://github.com/dart-lang/sdk/issues/50565 The native assets feature is available behind an experimental flag in Dart. It is not yet available in Flutter. https://github.com/flutter/flutter/issues/129757

**Comment by virtualzeta** [comment]
Thanks for your fast reply.

As often happens, an answer brings other doubts with it and, after having tried to resolve them on my own, I think it is appropriate that I ask you some questions (even basic ones) to dispel them.
I also hope it will help others like me who still have a lot to learn from Dart, Flutter and also from C.

I state that I am about to release an app and I would like to avoid changing SDK versions at this moment or prepare for the use of SDK manager, leaving the thing to other moments in which I can test the work environment more serenely .

I remember that at the moment I'm using this configuration: Flutter 3.7.3 • stable channel • Dart 2.19.2.

Let's see if these statements/questions == true. :-)

1. The use of the `--enable-experiment=native-asset`s flag of Dart is to be used only when launching/compiling/testing the Dart project concerned without it modifying the SDK in any way or remaining active subsequently for other projects ?
I tried first if the commands `--disable-experiment=native-assets` and `--disable-experiment responded` (as for `--disable-analytics`) but they are not recognized.

2. Waiting to have that eventual `globalVisibility` argument in the `DynamicLibrary.open` constructor in some release, I can now directly access the `dlopen` native function passing library name and Runtime Linker/Loader parameter depending on the platform.

3. The `dlopen` is a basic C function present in the `dlfnc.h` library and if in Dart I want to call it in the same way in the declaration below I omit the name as per the specifications of the Native class bearing in mind that in C it is declared as `void *dlopen(const char *filename, int flags)` (or I'd let ffigen do it if I'm starting from custom native code).

```
@Native<Pointer<Void> Function(Pointer<Char>, Int)>()
external Pointer<Void> dlopen(Pointer<Char> file, int mode);
```

4. My IDE (VS Code) doesn't give me any results for the `using`, `platformPath` and `dlopenPlatformSpecific` functions but above all I haven't found them in my SDK or in the last available one whose package I downloaded.
On the other hand I'm talking about functions and not about classes due to the camelCase used.
Initially I assumed they were external native functions in C proprietary to Dart and needed to be declared as done for dlopen, but the named parameter path made me think otherwise (at least for that function).
So are they special functions that I don't currently have and are only usable/recognized with Dart's experimental flag in runtime?

5. Is the `Native` class, which as I said I don't have available, usable/recognized as assumed above, therefore via the Dart flag?

6. As an alternative to having the original Native class available, can I declare one in the same file (or externally) as per the Dart spec like this? I see only one manufacturer and no restrictions/abstractions or anything, although I think there's a lot of cool stuff missing under the hood (and in this case it would be interesting to know where things stand).

```
class Native<T> {
   finalString? symbol;
   finalString? assets;
   final bool isLeaf;

   const Native({
     this.asset,
     this.isLeaf: false,
     this.symbol,
   });
}
```

7. Considering that the use made with `Native` gui does not use other parameters not present in `FfiNative`, can I still use that one using the non-omissable `nativeName`?

8. In the future, after the experimentation, will it be enough to use `@Native` with an `external` function declaration without worrying about whether the symbol is global or not because all the symbols of the loaded libraries will be present in the assets, without the need for them to be global or local?

9. Since I haven't had the chance to test the code yet due to what was written above, considering that I had no problems in Dart, once launched this will also work in Flutter even if for now the study is specific for Dart?

10. Are there too many questions? :-)

**Comment by virtualzeta** [comment]
> [#50105 (comment)](https://github.com/dart-lang/sdk/issues/50105#issuecomment-1441764283)
> 
> MacOS and Linux open with GLOBAL, while Android does not. That explains the difference between the Flutter and Dart experience. Flutter Desktop should also work with GLOBAL.
> 
> The workaround is to call `dlopen` yourself with the right arguments
> 
> ```dart
> /// On Linux and Android.
> const RTLD_LAZY = 0x00001;
> 
> /// On Android Arm.
> const RTLD_GLOBAL_android_arm32 = 0x00002;
> 
> /// On Linux and Android Arm64.
> const RTLD_GLOBAL_rest = 0x00100;
> 
> final RTLD_GLOBAL = Abi.current() == Abi.androidArm
>     ? RTLD_GLOBAL_android_arm32
>     : RTLD_GLOBAL_rest;
> 
> @Native<Pointer<Void> Function(Pointer<Char>, Int)>()
> external Pointer<Void> dlopen(Pointer<Char> file, int mode);
> 
> /// Returns dylib
> Object dlopenGlobalPlatformSpecific(String name, {String? path}) {
>   if (Platform.isLinux || Platform.isAndroid || Platform.isFuchsia) {
>     // TODO(https://dartbug.com/50105): enable dlopen global via package:ffi.
>     return using((arena) {
>       final dylibHandle = dlopen(
>           platformPath(name).toNativeUtf8(allocator: arena).cast(),
>           RTLD_LAZY | RTLD_GLOBAL);
>       return dylibHandle;
>     });
>   } else {
>     // The default behavior on these platforms is RLTD_GLOBAL already.
>     return dlopenPlatformSpecific(name, path: path);
>   }
> }
> ```
> 
> In the future, you will be able to use `@Native` without relying on opening things global. #50565 The native assets feature is available behind an experimental flag in Dart. It is not yet available in Flutter. [flutter/flutter#129757](https://github.com/flutter/flutter/issues/129757)

> [#50105 (comment)](https://github.com/dart-lang/sdk/issues/50105#issuecomment-1441764283)
> 
> MacOS and Linux open with GLOBAL, while Android does not. That explains the difference between the Flutter and Dart experience. Flutter Desktop should also work with GLOBAL.
> 
> The workaround is to call `dlopen` yourself with the right arguments
> 
> ```dart
> /// On Linux and Android.
> const RTLD_LAZY = 0x00001;
> 
> /// On Android Arm.
> const RTLD_GLOBAL_android_arm32 = 0x00002;
> 
> /// On Linux and Android Arm64.
> const RTLD_GLOBAL_rest = 0x00100;
> 
Thanks for your fast reply.

As often happens, an answer brings other doubts with it and, after having tried to resolve them on my own, I think it is appropriate that I ask you some questions (even basic ones) to dispel them.
I also hope it will help others like me who still have a lot to learn from Dart, Flutter and also from C.

I state that I am about to release an app and I would like to avoid changing SDK versions at this moment or prepare for the use of SDK manager, leaving the thing to other moments in which I can test the work environment more serenely .

I remember that at the moment I'm using this configuration: Flutter 3.7.3 • stable channel • Dart 2.19.2.

Here are some statements and questions for you to check.

1. What you said and linked to me made me understand that the difference in behavior is not related to the SDK but to the platform and, in this case, the question is really related to whether the symbols are local or global.
I presume that using Dart standalone the reference platform is in reference to the operating system used: I am using Windows and therefore have a the type of the RLTD is global. I also tried Flutter with Windows target platform and in fact I can use the example function because the symbols are global.

2. Waiting to have that eventual `globalVisibility` argument in the `DynamicLibrary.open` constructor in some release, I can now directly access the `dlopen` native function passing library name and Runtime Linker/Loader parameter depending on the platform.

4. The `dlopen` is a basic C function present in the `dlfnc.h` library and if in Dart I want to call it in the same way in the declaration below I omit the name as per the specifications of the `Native` class bearing in mind that in C it is declared as `void *dlopen(const char *filename, int flags)` (or I'd let ffigen do it if I'm starting from custom native code).

```
@Native<Pointer<Void> Function(Pointer<Char>, Int)>()
external Pointer<Void> dlopen(Pointer<Char> file, int mode);
```

5. The use of the `--enable-experiment=native-assets` flag of Dart is to be used only when launching/compiling/testing the Dart project concerned without it modifying the SDK in any way or remaining active subsequently for other projects?
I tried first if the commands `--disable-experiment=native-assets` and `--disable-experiment responded` (as for `--disable-analytics`) but they are not recognized.

6. My IDE (VS Code) doesn't give me any results for the `using`, `platformPath` and `dlopenPlatformSpecific`.
After unsuccessfully searching for them in my SDK and in the latest downloaded version for checking (unsuccessfully) I finally found `using` by importing `package:ffi/ffi.dart` to (unlike other used classes which stay in `dart:ffi`) and understood that the others were just personal functions to handle library name + path and native library fetched normally because the referenced platform already provides global symbols.
I used these, for test:
```
DynamicLibrary dlopenPlatformSpecific(String name, {String? path}) {
  DynamicLibrary nativeLib =
      DynamicLibrary.open(platformPath(name, path: path));
  return nativeLib;
}

String platformPath(String name, {String? path}) {
  path ??= "";
  return "$path$name";
}
```
7. Is the `Native` class, which as I said I don't have available, usable/recognized via the Dart flag?

8. As an alternative to having the original `Native` class available, I tried to declare one in the same file (or externally) as per Dart specifications like below, but obviously it is a class to implement and I would like to know where this is done for `Native` and `FfiNative` to know the code used.
```
class Native<T> {
   finalString? symbol;
   finalString? assets;
   final bool isLeaf;

   const Native({
     this.asset,
     this.isLeaf: false,
     this.symbol,
   });
}
```
The following error is returned:
```
I/flutter ( 6393): NoSuchMethodError: No top-level method 'dlopen' declared.
I/flutter ( 6393): Receiver: top-level
I/flutter ( 6393): Tried calling: dlopen()
```
9. Considering that the use made with `Native` here does not use other parameters not present in `FfiNative`, I tried to use that one using the non-omissable `nativeName` `'dlopen'` but although this does not throw an error with the library it returns with `address=0x0` and trying to use the native function then you get the error:
```
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: NoSuchMethodError: Class 'Pointer<Never>' has no instance method 'providesSymbol'.
E/flutter ( 5444): Receiver: Pointer: address=0x0
E/flutter ( 5444): Tried calling: providesSymbol("add")
```
10. In the future, after the experimentation, will it be enough to use `@Native` with an `external` function declaration without worrying about whether the symbol is global or not because all the symbols of the loaded libraries will be present in the assets, without the need for them to be global or local?

I look forward to your instructions in order to successfully start the script.

**Comment by virtualzeta** [comment]
> I was thinking that the return of the handle with zeroed address can somehow also be connected to an incorrect path. But since I'm sure the path to the library is correct for Android (I just put no path but just the standard name of the library lib<library_name>.so) and I can read the function with DynamicLibrary.open constructor and lookup, is it possible that using it adds a part of the path that you need to add manually with dlopen instead?
> The error with Windows leaves me a bit perplexed but I'm inclined towards a path error.
> I tried to add via path_provider package (with trailing slash):
> 
> /data/user/0/com.example.<app_name>/app_flutter
> /data/user/0/com.example.<app_name>/cache
> /data/user/0/com.example.<app_name>/files
> Although Directory(path).existsSync() is true, File(platformPath(name, path: path)).existsSync() is always false.
> And finally I found the correct path is /data/user/0/com.example.<app_name>/lib/lib<library_name>.so, manually adding "/lib" because getLibraryDirectory() is not compatible with Android.
> 
> Too bad that even with the correct path to pass to dlopen native function, the handler returns with zeroed address. >:-|

Regarding point 9 I opened a new problem to analyze it individually and even if I understood the main problem I also have to understand how to solve it.

[https://github.com/dart-lang/ffi/issues/204](https://github.com/dart-lang/ffi/issues/204)

**Comment by dcharkes** [comment]
> 4\. My IDE (VS Code) doesn't give me any results for the `using`, `platformPath` and `dlopenPlatformSpecific`.

`using` is in `package:ffi`.

The other two you'll have to define yourself, for your convenience:

```dart
String platformPath(String name, {String path = ""}) {
  if (Platform.isLinux || Platform.isAndroid || Platform.isFuchsia)
    return path + "lib" + name + ".so";
  if (Platform.isMacOS) return path + "lib" + name + ".dylib";
  if (Platform.isWindows) return path + name + ".dll";
  throw Exception("Platform not implemented");
}
```

> 5\. The use of the `--enable-experiment=native-assets` flag of Dart is to be used only when launching/compiling/testing the Dart project concerned without it modifying the SDK in any way or remaining active subsequently for other projects?

Only in the one invocation. (Dart experiments are per invocation. Flutter experiments are global and persisted. So once native assets are available there it will be a persisted setting.)

> 6\. Is the `Native` class, which as I said I don't have available, usable/recognized via the Dart flag?

It's available `@Since('2.19')`. Are you using an older Dart?

> But although in Android this does not throw an error with the dlopen it returns an handle with `address=0x0` and I certainly can't fetch symbols.

If you can call it, you should be able to use it. Maybe you're passing the wrong path to the dynamic library or the wrong flags.

> But with the Windows platform (which shouldn't be affected given the premises but I used it as a test) the behavior is different and directly returns an error.

Windows has a different API and does not use `dlopen` at all. When you use `DynamicLibrary.open` in Dart on a Windows machine it calls the corresponding Windows API instead. You should not have to do this on Windows, because by default everything is opened globally.

**Comment by virtualzeta** [comment]
> using is in package:ffi.

Yes, as written previously I used `import 'package:ffi/ffi.dart'`. Alone `package:ffi` does not find the path.

> It's available @Since('2.19'). Are you using an older Dart?

Yes, as stated and indicated in my comments, for a correct analysis, I use Dart 2.19.2.

In my version @FfiNative is already indicated as deprecated BUT `@Native` is not available as I have seen in the latest version (and certainly in other previous ones). I suspect there was some botch in the release.

Anyway this in my case shouldn't make any difference because the native function is also found with `@FfiNative` (as well as using `lookup`) and the problem is that since they are not the global symbols are unknown. This as a system of what is indicated more extensively in the previous comments and in the problem opened separately following numerous tests.

**Comment by dcharkes** [comment]
Correct, `@FfiNative` and `@Native` should be identical in functionality. It's just a different syntax.
